sdk/core/azure-core/inc/http/http.hpp

```diff
@@ -3,27 +3,89 @@
 
 #pragma once
 
+#include <string>
+#include <vector>
+
 #include <internal/contract.hpp>
 
+using namespace std;
+
 namespace azure
 {
 namespace core
 {
 namespace http
 {
 
-class http_request
+namespace http_method
 {
+enum HttpMethod
+{
+  GET,
+  HEAD,
+  POST,
+  PUT,
+  DELETE,
+  PATCH,
+};
+}
 
+class Header
+{
 private:
-  /* data */
+  string name;
+  string value;
+
 public:
-  http_request();
-  ~http_request();
+  Header(string name, string value)
+  {
+    this->name = name;
+    this->value = value;
+  }
+  ~Header() {}
+  string getName();
+  string getValue();
 };
 
-http_request::http_request() {}
-http_request::~http_request() {}
+class Request
+{
+
+private:
+  http_method::HttpMethod method;
+  string url;
+  std::vector<Header> headers;
+  string body;
+  int32_t query_start; // 0 = no query in url, > 0 = query start position in url '?'
+
+  inline int32_t get_query_start(string url)
+  {
+    auto position = url.find('?');
+    return position == string::npos ? 0 : position;
+  }
+
+public:
+  Request(http_method::HttpMethod httpMethod, string url)
```

Comment:
std::string const& url

Reply:
really?  pointer ?  why? 
Doesn't it make sense to create a copy of the url?   or do we want to support one url for multiple requests?

Reply:
One thing Storage does is to make sure each member variable that copies the input actually uses pass-by-value, then use `std::move` inside the ctor/function. This avoids double copy for rvalues.

Reply:
Normally this applies to setters/ctors.

Reply:
@vhvb1989 not a pointer but reference. When you pass a string (X) that you have in your function by value, it another (temporary) instance of std::string (Y) may get created. Which also may involve copying string contents. All that to assign it to another string (data member) (Z) which is the final destination, and once again memory contents are getting copied (or at least  behaves as if memory contents were copied.)
If you were to declare Z as a reference then you likely shouldn't do that. But for function arguments, in C++, we should almost always pass classes/structs by const reference.

Reply:
> Normally this applies to setters/ctors.

I agree with @katmsft on this matter.

Consider we're passing an rvalue `url` to the constructor, if the constructor accepts const reference, it's going to make a copy. But if it accepts by value, we can directly move the rvalue into member variable.

Reply:
updated all strings parameters to use refs